Implement TripOptions for queryTrips method

ialokim opened the pull request:
This PR moves the following "auxiliary" parameters from `queryTrips` into a separate `TripOptions` class while leaving `from`, `via`, `to`, `date` and `dep` as normal parameters: `products`, `optimize`, `walkSpeed`, `accessibility` and `options`.
See also the quick discussion in #214.

For this, I've added a new method signature in the `NetworkProvider` interface while marking the old one as _deprecated_. I've also changed the implementations in all the subclasses to use the new parameters and implemented the old method signature in `AbstractNetworkProvider` as a simple alias for the new one.

I have run all the live tests to make sure the PR does not break any providers. There were exactly the same amount of failures before and after the changes, but I was not able to test providers requiring secret keys except the Navitia based ones.

Fixes #214.

schildbach commented on enabler/src/de/schildbach/pte/NetworkProvider.java:
```diff
@@ -68,6 +68,38 @@
     }
 
+    public class TripOptions {
+        public @Nullable Set<Product> products;
```
Can we make these fields final?

ialokim replied:
Sure, but then we would have to change some logic where e.g. `options.products` is excepted to be non-null (see [here](https://github.com/schildbach/public-transport-enabler/pull/216/files/13e16d3e4571d96572828b2388dfaeb058aefe6e#diff-b2543011aff601f59664df7f7e17430cR851) and some lines downwards as an example).

Do you want me to amend to this commit or add a new commit with these changes or do you want to change it while merging?

schildbach replied:
Oh yes, indeed. Please do not mutate the TripOptions instance. It's owned by the caller, who will not expect it to be changed as a side effect of calling.

schildbach commented on enabler/src/de/schildbach/pte/NetworkProvider.java:
```diff
@@ -160,8 +188,8 @@ QueryTripsResult queryTrips(Location from, @Nullable Location via, Location to,
      * 
      * @param context
      *            context to query more trips from
-     * @param next
```
This fix seems unrelated to the PR. Please don't mix.

ialokim replied:
Ok, I'll change that.

schildbach replied:
Yes, please do. You can fix it in a separate PR if you want, then I'll merge the JavaDoc fix right away.

schildbach commented on enabler/src/de/schildbach/pte/VrsProvider.java:
```diff
@@ -688,6 +686,11 @@ public QueryTripsResult queryTrips(final Location from, final @Nullable Location
 
+        if (options == null)
+            options = new TripOptions();
+        if (options.products == null)
```
These two lines also seem to be unrelated, but let me know if I'm wrong.

ialokim replied:
Hum they are kind of related, because in line 700 (old file) it expected the `products` to be non-null, so that's what I wanted to ensure here.

schildbach replied:
I understand the problem, and I agree the NPE should be fixed. It's just not related to this PR, as the bug has existed before.

I just fixed the bug in master via 458fea99ac0bbe4dc8a526bb9e98d8c4f3eabb88. If you rebase your PR on master the need for this specific change should go away.

grote commented:
> I wonder if we perhaps should rename the current Options enum to TripFlags to make the API clearer.

Good idea!

schildbach commented:
Then again if we plan the Options to TripFlags rename in this case the API will break anyway so there would be not much point backwards compatibility in this PR.

ialokim commented:
Thanks @schildbach for your comments.

I've now rebased the PR on current master, changed the `TripOptions` fields to final (I've also had to change the `products` handling in the `NegentweeProvider.java`) and split out the javadoc fix into #217.

schildbach commented:
I merged this, many thanks! I decided to move TripOptions to the .dto package.
